### DISCUSSION TITLE
Hide mouse cursor on a trial

### DISCUSSION THREAD

**Discussion by grocio** [jspsych/jsPsych#1407] [Q&A] [answered]
I'd like to hide the mouse cursor and allow keyboard input at a trial-level (surprisingly, hiding the cursor disables keyboard input in a certain case). I know a solution but I'd like to know best practices.

In a relevant thread on jsPsych's google group [https://groups.google.com/g/jspsych/c/edptYoa0JBc](https://groups.google.com/g/jspsych/c/edptYoa0JBc), @jodeleeuw recommends using the CSS style. So, I tried 
```javascript
var cursor_off = {
    type: 'call-function',
    func: function() {
        document.body.style.cursor= "none";
    }
}

var cursor_on = {
    type: 'call-function',
    func: function() {
        document.body.style.cursor= "auto";
    }
}

timeline.push(cursor_off, some_trial, cursor_on);
```
It made cursor invisible/visible. **It, however, disabled keyboard input (i.e., html-keyboard-response in my case).** I assume that it is because {cursor: "none";} in body makes the web page inactive (not focused).

Then, I used a new feature (#832; custom CSS class per trial). Specifically, I added
```javascript
<style>
.hide_cursor {
    cursor: none;
}
</style>
```
and I applied this CSS style to a trial

```javascript
var trial = {
    type: 'html-keyboard-response',
    stimulus: '331',
    choices: ['j','k','l', 'J', 'K', 'L'],
    css_classes: ['hide_cursor']
};
```

It hided the mouse cursor only if it is on the stimulus. The region of the stimuli (e.g., '331') was small. So, I increased its region by using padding in the CSS style (`padding: 30%`).

This worked well. I show an example below,
```javascript
... someting ...

<style>
.hide_cursor {
    cursor: none;
}
.stimulus_size {
    padding: 30%;
}
</style>

... something ...

var trial = {
    type: 'html-keyboard-response',
    stimulus: '331',
    choices: ['j','k','l', 'J', 'K', 'L'],
    css_classes: ['hide_cursor', 'stimulus_size']
};
```
It worked well as I intended and this is a tentative solution. I wonder if you could suggest a more elegant solution?

**Comment by jodeleeuw** [comment]
I'm converting this to a discussion post.

**Comment by jodeleeuw** [comment, marked as answer]
Hi @grocio,

> It made cursor invisible/visible. It, however, disabled keyboard input (i.e., html-keyboard-response in my case). I assume that it is because {cursor: "none";} in body makes the web page inactive (not focused).

This seems odd to me. I don't think `{cursor: "none";}` should make the page inactive. I created a demo here:

https://1lpsifxdjx.cognition.run/

The code is:

```js
var first = {
  type: 'html-keyboard-response',
  stimulus: `<p>First trial just to compare. Cursor visible.</p>
    <p>Press any key.</p>`
}

var cursor_off = {
    type: 'call-function',
    func: function() {
        document.body.style.cursor= "none";
    }
}

var second = {
  type: 'html-keyboard-response',
  stimulus: `<p>Second trial. Cursor off.</p>
    <p>Press any key.</p>`
} 

var cursor_on = {
    type: 'call-function',
    func: function() {
        document.body.style.cursor= "auto";
    }
}

var third = {
  type: 'html-keyboard-response',
  stimulus: `<p>Third trial. Cursor on.</p>
    <p>Press any key.</p>`
} 

jsPsych.init({timeline: [first, cursor_off, second, cursor_on, third]})
```

This works OK for me. I wonder if you didn't have focus on the window when you hid the cursor?

**Reply by grocio** [reply]
Thank you for your quick response. I also appreciate your demo. Your demo works perfectly!

I made an elementary mistake in my previous script: I forgot that jsPsych records everything and trials of call-function are saved in `jsPsych.data` (as row entries in a csv file). As I used `jsPsych.data` in looping and conditional timelines, I had to refer to the correct trial. For example, when referring to data of trial A in sequence of "trial A -> call-function trial (e.g., cursor_on)", I had to data of the second-to-last trial, skipping the call-function trial.

Once again, thank you very much. I used the above mentioned cursor on/off trick in my pilot experiment. The results look perfect.

**Comment by donknow** [comment]
Hello, 
I tried the above-mentioned solution, but the type was changed since then and now asks for a plugin. What is the corresponding plugin called?
Best wishes

**Reply by becky-gilbert** [reply]
Hi @donknow, are you using jsPsych v7? If so you can follow the [v6 to v7 migration guide](https://www.jspsych.org/7.3/support/migration-v7/) to change the code. 

But to answer your question, the `type` parameter has changed from a string to a class name in v7, so the trial objects would look like this:
```js
var first = {
  type: jsPsychHtmlKeyboardResponse,
  stimulus: `<p>First trial just to compare. Cursor visible.</p>
    <p>Press any key.</p>`
}

var cursor_off = {
    type: jsPsychCallFunction,
    func: function() {
        document.body.style.cursor= "none";
    }
}

var second = {
  type: jsPsychHtmlKeyboardResponse,
  stimulus: `<p>Second trial. Cursor off.</p>
    <p>Press any key.</p>`
} 

var cursor_on = {
    type: jsPsychCallFunction,
    func: function() {
        document.body.style.cursor= "auto";
    }
}

var third = {
  type: jsPsychHtmlKeyboardResponse,
  stimulus: `<p>Third trial. Cursor on.</p>
    <p>Press any key.</p>`
} 
```
